lib/SyTest/Federation/Datastore.pm

```diff
@@ -200,6 +200,64 @@ sub create_event
    return $event;
 }
 
+=head2 get_backfill_events
+
+   @events = $store->get_backfill_events( start_at => \@ids, ... )
+
+Returns a list of events, starting from the event(s) whose ID(s) are given by
+the C<start_at> argument, and continuing backwards through their
+C<prev_events> linkage. They are returned in a linearized order, most recent
+first.
+
+The following other named arguments affect the behaviour:
+
+=over 4
+
+=item limit => $limit
+
+Gives the maximum number of events that should be returned.
+
+=item stop_before => \@ids
+
+Gives a list of event IDs that should not be entered into or returned (most
+likely because the caller already has them). These events do not count towards
+the overall count limit.
+
+=back
+
+=cut
+
+sub get_backfill_events
+{
+   my $self = shift;
+   my %params = @_;
+
+   my $start_at = $params{start_at} or
+      croak "Require 'start_at'";
+
+   my $limit = $params{limit} or
+      croak "Require 'limit'";
+
+   my @event_ids = @$start_at;
+
+   my %stop_before = map { $_ => 1 } @{ $params{stop_before} // [] };
+
+   my @events;
+   while( @event_ids and @events < $limit ) {
+      my $id = shift @event_ids;
+      my $event = eval { $self->get_event( $id ) }
+         or next;
+
+      push @events, $event;
+
+      push @event_ids, grep { !$stop_before{$_} }
+                       map { $_->[0] } @{ $event->{prev_events} };
+      $stop_before{$id}++;
```

Comment:
Perhaps change this to:

```perl
my %exclude = map { $_ => 1 } @{ $params{stop_before} // [] };
```
because you're changing the meaning of the stop_before hash to really be a set of things not to include.

And here, $exclude{$id} = 1; rather than ++, because your semantics here are adding to the set (boolean) not counting.